### PR TITLE
Adding Azure SignalR provisioning and a showcase to playground

### LINKED ISSUE #799: Add Azure SignalR Aspire Component
We should add an Azure SignalR Component. See https://learn.microsoft.com/azure/azure-signalr/

See https://www.nuget.org/packages/Microsoft.Azure.SignalR

### PR DESCRIPTION
**Pull request by vicancy**
Fixes https://github.com/dotnet/aspire/issues/799

* Add the provisioning part of Azure SignalR into aspire
* Azure SignalR SDK will support reading connection name with its `AddNamedAzureSignalR(string connectionName)`, the API naming could be changed if there are any better options
* Add a signalr playground sample following SignalR quick start tutorial https://learn.microsoft.com/aspnet/core/tutorials/signalr?view=aspnetcore-8.0&tabs=visual-studio-code#add-signalr-client-code with minimum change: adding  `AddNamedAzureSignalR("signalr")`
* Add bicep for signalr
** Sample provisioning output **
image

** Sample SignalR web server**
image

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1873)

### CODE REVIEW

**Comment by davidfowl** [comment]
@vicancy Can you add a new playground app for the SignalR service sample instead of modifying orleans?

**Comment by vicancy** [comment]
Hi @eerhardt could you help add https://www.myget.org/feed/azure-signalr-dev/package/nuget/Microsoft.Azure.SignalR/1.25.0-preview1-11147 to the nuget source? This is the preview package containing the newly added API `AddAzureSignalRWithConnectionName`. Also could you share what the process is when the dependent package version is updated? Thanks!

File: playground/signalr/SignalRWeb/wwwroot/js/chat.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+connection.on("ReceiveMessage", function (user, message) {
+    var li = document.createElement("li");
+    document.getElementById("messagesList").appendChild(li);
+    // We can assign user-supplied strings to an element's textContent because it
+    // is not interpreted as markup. If you're assigning in any other way, you 
+    // should be aware of possible script injection concerns.
+    li.textContent = `${user} says ${message}`;
+});
+
+connection.start().then(function () {
```
Comment (davidfowl) [inline playground/signalr/SignalRWeb/wwwroot/js/chat.js:17]: no async/await?

**Comment by eerhardt** [comment]
> Hi @eerhardt could you help add https://www.myget.org/feed/azure-signalr-dev/package/nuget/Microsoft.Azure.SignalR/1.25.0-preview1-11147 to the nuget source?

I don't believe we can mirror a package from myget.org. Can we use a version that is on nuget.org?

> Also could you share what the process is when the dependent package version is updated?

Checkout https://github.com/dotnet/arcade/blob/main/Documentation/MirroringPackages.md

> If the package is owned by Microsoft, new versions (or versions not originally mirrored), will be automatically updated the next time the pipeline runs on its schedule.

File: playground/signalr/SignalRWeb/wwwroot/js/chat.js (hunk 1 above)
Comment (vicancy) [reply playground/signalr/SignalRWeb/wwwroot/js/chat.js:17]: Exactly following the code in https://learn.microsoft.com/en-us/aspnet/core/tutorials/signalr?view=aspnetcore-8.0&tabs=visual-studio-code#add-signalr-client-code :D

**Comment by vicancy** [comment]
> > Hi @eerhardt could you help add https://www.myget.org/feed/azure-signalr-dev/package/nuget/Microsoft.Azure.SignalR/1.25.0-preview1-11147 to the nuget source?
> 
> I don't believe we can mirror a package from myget.org. Can we use a version that is on nuget.org?
> 
> > Also could you share what the process is when the dependent package version is updated?
> 
> Checkout https://github.com/dotnet/arcade/blob/main/Documentation/MirroringPackages.md
> 
> > If the package is owned by Microsoft, new versions (or versions not originally mirrored), will be automatically updated the next time the pipeline runs on its schedule.

Thanks for the explanation. Updated to nuget: https://www.nuget.org/packages/Microsoft.Azure.SignalR/1.25.0-preview1-11147

**Comment by davidfowl** [comment]
@vicancy we're swapping azure resources over to the [new bicep model](https://github.com/dotnet/aspire/tree/main/src/Aspire.Hosting.Azure/Bicep). This means you don't need to update azd to enable deploying it.

**Comment by vicancy** [comment]
Hi @eerhardt also help add Microsoft.Azure.SignalR.Protocols package please

File: src/Aspire.Hosting.Azure.Provisioning/Provisioners/AzureSignalRProvisioner.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure.Provisioning/Provisioners/AzureSignalRProvisioner.cs:1]: You can delete this and the associated code.

**Comment by eerhardt** [comment]
> Hi @eerhardt also help add Microsoft.Azure.SignalR.Protocols package please

I've mirrored https://github.com/dotnet/aspire/pull/1873#issuecomment-1955956313 to the dotnet-public feed. It will also mirror all its dependencies as well.

**Comment by vicancy** [comment]
Thanks for the help @eerhardt , we released a stable version 1.25.0, hopefully the pipeline will automatically mirror the package right?

The PR is ready for review, @eerhardt @davidfowl

**Comment by eerhardt** [comment]
Can the playground sample be simplified? It seems like a lot of css and js files being checked in for little value. I know this came from a template, but it just feels like more than what is necessary.

image

**Comment by vicancy** [comment]
@eerhardt Thanks for the feedback, makes total sense, updated.

**Comment by davidfowl** [comment]
Looks really good!

File: src/Aspire.Hosting.Azure/Bicep/signalr.bicep [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+param tags object = {}
+
+@description('The pricing tier of the SignalR resource.')
+@allowed([
+  'Free_F1'
+  'Standard_S1'
+  'Premium_P1'
+])
+
+param pricingTier string = 'Free_F1'
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/Bicep/signalr.bicep:14]: should we call this sku?

**Comment by davidfowl** [comment]
@eerhardt This change looks really good, I'm willing to deal with any fallout for P4 (can merge tomorrow).

File: src/Aspire.Hosting.Azure/Bicep/signalr.bicep (hunk 3 above)
Comment (vicancy) [reply src/Aspire.Hosting.Azure/Bicep/signalr.bicep:14]: sku contains both unit number and pricing tier  
image

Comment (davidfowl) [reply src/Aspire.Hosting.Azure/Bicep/signalr.bicep:14]: Got it.

**Review by eerhardt — APPROVED** [review]
Thanks for the contribution, @vicancy!

I think we can merge this today.

**Pull request merged**